extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeSignatureUtils.java

```diff
@@ -0,0 +1,150 @@
+     */
+    @Nonnull
+    static String getDeobfuscationCode(@Nonnull final String javaScriptPlayerCode)
+            throws ParsingException {
+        try {
+            final String deobfuscationFunctionName = getDeobfuscationFunctionName(
+                    javaScriptPlayerCode);
+
+            String deobfuscationFunction;
+            try {
+                deobfuscationFunction = getDeobfuscateFunctionWithLexer(
+                        javaScriptPlayerCode, deobfuscationFunctionName);
+            } catch (final Exception e) {
+                deobfuscationFunction = getDeobfuscateFunctionWithRegex(
+                        javaScriptPlayerCode, deobfuscationFunctionName);
+            }
+
+            // Assert the extracted deobfuscation function is valid
+            JavaScript.compileOrThrow(deobfuscationFunction);
+
+            final String helperObjectName =
+                    Parser.matchGroup1(SIG_DEOBF_HELPER_OBJ_NAME_REGEX, deobfuscationFunction);
+
+            final String helperObject = getHelperObject(javaScriptPlayerCode, helperObjectName);
+
+            final String callerFunction = "function " + DEOBFUSCATION_FUNCTION_NAME
+                    + "(a){return "
+                    + deobfuscationFunctionName
+                    + "(a);}";
+
+            return helperObject + deobfuscationFunction + callerFunction;
+        } catch (final Exception e) {
+            throw new ParsingException("Could not parse deobfuscation function", e);
+        }
+    }
+
+    @Nonnull
+    private static String getDeobfuscationFunctionName(@Nonnull final String javaScriptPlayerCode)
+            throws ParsingException {
+        Parser.RegexException exception = null;
+        for (final String regex : FUNCTION_REGEXES) {
+            try {
+                return Parser.matchGroup1(regex, javaScriptPlayerCode);
+            } catch (final Parser.RegexException e) {
+                if (exception == null) {
+                    exception = e;
+                }
+            }
+        }
+
+        throw new ParsingException(
+                "Could not find deobfuscation function with any of the known patterns", exception);
+    }
+
+    @Nonnull
+    private static String getDeobfuscateFunctionWithLexer(
+            @Nonnull final String javaScriptPlayerCode,
+            @Nonnull final String deobfuscationFunctionName) throws ParsingException {
+        final String functionBase = deobfuscationFunctionName + "=function";
+        return functionBase + JavaScriptExtractor.matchToClosingBrace(
+                javaScriptPlayerCode, functionBase) + ";";
+    }
+
+    @Nonnull
+    private static String getDeobfuscateFunctionWithRegex(
+            @Nonnull final String javaScriptPlayerCode,
+            @Nonnull final String deobfuscationFunctionName) throws ParsingException {
+        final String functionPattern = DEOBF_FUNC_REGEX_START
+                + deobfuscationFunctionName.replace("$", "\\$")
```

Comment:
Here and in `getHelperObject` you used `.replace($, \$)`, while in `YoutubeThrottlingParameterUtils.getDeobfuscationFunctionName` you used `Pattern.quote()`. Is there a reason for this? I think `replace` is just fine, assuming that the variable name contains just `a-zA-Z0-9_$`, which is what the regexes enforce. (in theory js variables can be [as stupid as you want](https://stackoverflow.com/q/1661197))

Reply:
The `.replace($, \$)` calls have been not written by me, they have been introduced in the initial commit of the extractor. I don't know the reason.

I would personally use `Pattern.quote` to avoid any issues in the future, in the case YouTube starts to use weird names.